### DISCUSSION TITLE
`payload generate:importmap` Fails with Custom Lexical Editor Feature

### DISCUSSION THREAD

**Discussion by Tanish2002** [payloadcms/payload#10011] [Q&A] [answered]
### Describe the Bug

I've implemented a custom feature for the Lexical editor that allows marking text with colors(just add a tailwind class). This feature is heavily inspired by the `LinkFeature` code used by payload.

While the feature is nearly complete, I encounter an error when running the command `payload generate:importmap`. Below is the error output:
```
❯ pnpm payload generate:importmap                                                                                                                                                                            took 4m12s 

> custom-color-feature-repro@1.0.0 payload /home/weeb/code/javascript/custom-color-feature-repro
> cross-env NODE_OPTIONS=--no-deprecation payload "generate:importmap"


node:internal/process/promises:391
    triggerUncaughtException(err, true /* fromPromise */);
    ^
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".css" for /home/weeb/code/javascript/custom-color-feature-repro/node_modules/.pnpm/react-image-crop@10.1.8_react@19.0.0/node_modules/react-image-crop/dist/ReactCrop.css
    at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:176:9)
    at defaultGetFormat (node:internal/modules/esm/get_format:219:36)
    at defaultLoad (node:internal/modules/esm/load:143:22)
    at async nextLoad (node:internal/modules/esm/hooks:868:22)
    at async load (file:///home/weeb/code/javascript/custom-color-feature-repro/node_modules/.pnpm/tsx@4.19.2/node_modules/tsx/dist/esm/index.mjs?1734331094830:2:1762)
    at async nextLoad (node:internal/modules/esm/hooks:868:22)
    at async Hooks.load (node:internal/modules/esm/hooks:451:20)
    at async handleMessage (node:internal/modules/esm/worker:196:18) {
  code: 'ERR_UNKNOWN_FILE_EXTENSION'
}

Node.js v20.18.1
 ELIFECYCLE  Command failed with exit code 1.
```

## My Own Observations
Through testing, I narrowed the issue to the `createNode` method in the server-side feature code. The problem arises specifically with the nodes array that includes the following two custom nodes:
- `AutoColorTextNode`
- `ColorTextNode`

If I remove the `nodes` array from the `createServerFeature` implementation, the `payload generate:importmap` command works correctly without any errors.

---

If I remove the nodes array, run `payload generate:importmap` then run the dev server `pnpm run dev` and then simply re-add the nodes array it somehow automatically works, while the `payload generate:importmap` still fails. Below is an image of it working in the admin panel:
![Image](https://github.com/user-attachments/assets/8bca4331-7d2a-4b49-9e56-28b52ee08114)

Since it works using the above method, I can't figure out why this is happening,

### Link to the code that reproduces this issue

https://github.com/Tanish2002/custom-color-feature-repro

## Error Scenario
1. Clone the reproduction repository:
```bash
git clone <repository-link>
```
2. Navigate to the cloned repository and download dependencies:
```bash
cd custom-color-feature-repro
pnpm install
```
3. Run the following command:
```bash
pnpm payload generate:importmap
```
4.Observe the error output, which matches the following:
```bash
TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".css" ...
```

## Workaround

To bypass the issue and proceed with development:

1. Clone the reproduction repository and install dependencies (if not already done):
```bash
git clone https://github.com/Tanish2002/custom-color-feature-repro
cd custom-color-feature-repro
pnpm install
```
2. Open the file `src/features/colorText/server/index.ts`.
3. Locate the `createServerFeature` function and remove the nodes array from the returned object.
For example, change:
```ts
return {
  ...
  nodes: [
    createNode({ ... }),
    createNode({ ... }),
  ],
  ...
}
```
To:
```ts
return {
  ...
  // nodes array removed
}
```
4. Run the following commands:
```bash
pnpm payload generate:importmap
pnpm run dev
```
5. After the app starts, re-add the nodes array in `createServerFeature` to restore its functionality.
6. Allow the application to hot-reload and verify the feature is working as expected in the admin panel.

If you re-run the `pnpm payload generate:importmap` command after re-adding the nodes array, the same error occurs.

### Which area(s) are affected? (Select all that apply)

plugin: other, plugin: richtext-lexical

### Environment Info

```text
Binaries:
  Node: 20.18.1
  npm: 10.8.2
  Yarn: N/A
  pnpm: 9.15.0
Relevant Packages:
  payload: 3.7.0
  next: 15.1.0
  @payloadcms/email-nodemailer: 3.7.0
  @payloadcms/graphql: 3.7.0
  @payloadcms/next/utilities: 3.7.0
  @payloadcms/payload-cloud: 3.7.0
  @payloadcms/richtext-lexical: 3.7.0
  @payloadcms/translations: 3.7.0
  @payloadcms/ui/shared: 3.7.0
  react: 19.0.0
  react-dom: 19.0.0
Operating System:
  Platform: linux
  Arch: x64
  Version: #1-NixOS SMP PREEMPT_DYNAMIC Fri Nov 22 14:38:37 UTC 2024
  Available memory (MB): 15729
  Available CPU cores: 12
```

**Comment by rilrom** [comment]
That error usually implies there is something wrong with your path (in this case, ClientFeature in createServerFeature).

However I've only seen it in collection level imports, not custom lexical features so I could be wrong.

One thing I can see is that your client feature is using an index.tsx, whereas the example is using index.ts. Could be nothing but worth a shot.

**Comment by Tanish2002** [comment]
> That error usually implies there is something wrong with your path (in this case, ClientFeature in createServerFeature).

I've tried all types of path formats, converting the import to a default import using `#default` or `#ComponentName`, including full paths with `index.tsx#ComponentName/default`

> One thing I can see is that your client feature is using an index.tsx, whereas the example is using index.ts. Could be nothing but worth a shot.

Tried that, didn't work as well, the only reason I'm using `.tsx` is because of the `ChildComponent` property is supposed to be a React Component

**Comment by Tanish2002** [comment]
@rilrom One more thing, I don't think this is even caused by the `ClientFeature` Import, Like I previously mentioned, the error goes away if I remove the `nodes` array while the `ClientFeature` import is still there. 

If I remove the `nodes` array and then run `payload generate:importmap` it works fine which I can even confirm in the generated `importMap.js` has the ClientFeature mapped properly.

So I believe the issue lies somewhere in `nodes` array 🤔. However not confirmed

**Comment by rilrom** [comment]
So it looks to be specific to how you have implemented your custom nodes.

If I replace your custom node with the below:

```
export class ColorTextNode extends ElementNode {
  static getType(): string {
    return 'colorText'
  }
}
```

Then `importmap` works as expected.

I recommend slowly working your way through one of your custom nodes until you find the cause.

**Comment by Tanish2002** [comment]
Thanks, @rilrom!

I’ve identified the source of the bug in the `ColorTextNode` implementation, specifically in the following function:
```
function $convertAnchorElement(domNode: Node): DOMConversionOutput {
  let node: ColorTextNode | null = null;
  if (isHTMLElement(domNode)) {
    const content = domNode.textContent;
    if (content !== null && content !== '') {
      node = $createColorTextNode({
        id: new ObjectID().toHexString(),
        fields: {
          textColor: domNode.className,
        },
      });
    }
  }
  return { node };
}
```
The issue seems to be with the `isHTMLElement(domNode)` check. This utility is imported from `@payloadcms/richtext-lexical/client` and is also used in their `LinkFeature` implementation (via `isHTMLAnchorElement`, which internally uses `isHTMLElement`). However, I’m puzzled about why this is causing the problem in this context.

As a workaround, I’ve written a custom helper function to specifically check for span elements, which resolves the issue for now:
```
function $convertAnchorElement(domNode: Node): DOMConversionOutput {
  if (isSpanElement(domNode)) {
    const content = domNode.textContent;
    if (content) {
      const node = $createColorTextNode({
        id: new ObjectID().toHexString(),
        fields: {
          textColor: domNode.className,
        },
      });
      return { node };
    }
  }
  return { node: null };
}

function isSpanElement(domNode: Node): domNode is HTMLSpanElement {
  return domNode.nodeType === 1 && (domNode as HTMLElement).tagName.toLowerCase() === "span";
}
```
Would appreciate any insights into why `isHTMLElement` is behaving this way.. So gonna keep this issue open for now as this might be a bug either with the lexical package or payload 🤔

**Comment by rilrom** [comment]
Glad you've found a workaround!

Could you try importing it from '@payloadcms/richtext-lexical/lexical/utils' instead?

It seems the isHTMLElement that you are referencing is different to the one used in the link feature.

**Comment by Tanish2002** [comment]
That works! But why? It's the exact same [export](https://github.com/payloadcms/payload/blob/00909ec5c450a70a8293c75b9bd5bd12a794b97e/packages/richtext-lexical/src/exports/client/index.ts#L57) and [here](https://github.com/payloadcms/payload/blob/00909ec5c450a70a8293c75b9bd5bd12a794b97e/packages/richtext-lexical/src/lexical/utils/guard.ts)

**Comment by Tanish2002** [comment]
> It seems the isHTMLElement that you are referencing is different to the one used in the link feature.

Yeah I just realized that as well, but that still doesn't explain the issue with importMap command

**Comment by jmikrut** [comment, marked as answer]
Hey @Tanish2002 I can shine some light on this. You are importing a CSS file in server-only code, which is crashing the `generate:importmap` command (because that is a Node command and doesn't know how to parse CSS).

This is why all of our CSS for Lexical features is abstracted out of the server-side files and is present only in the files meant to be consumed solely in a browser context. Browser code goes through Turbopack / Webpack, but the config itself (and the server-side part of Lexical components) is Node-only, so there, CSS will break the process.

If you have a barrel file, or a single file that imports / exports many things (including a CSS file), this would crash the Node scripts that Payload uses. The `@payloadcms/richtext-lexical/client` export of ours is meant to be imported _**only in client-side code**_ - not Node code. That's why if you import that utility directly, it doesn't crash, but if you import from our client-side export, other CSS in that file will cause it to crash.

Does that make sense? I'll convert this to a discussion because I think that it will be beneficial for others. Happy to continue to help as well.

**Comment by Tanish2002** [comment]
@jmikrut That makes a lot more sense now! Thanks.

I do think the docs should mention that Nodes shouldn't contain any client side code. Also `createNode` could also mention this in it's jsdoc. 
Lastly the `isHTMLElement` should also be re-exported from `lexical/utils`since `isHTMLAnchorElement` is re-exported, maybe with a different name `isNodeHTMLElement` or rename the existing one.

**Comment by revnelson** [comment]
I'm getting this same error when using `const { data, rowNumber } = useRowLabel<{ title?: string }>()` in my custom component. When I remove that line, generating the import map works fine. With that line, I get the ReactCrop unknown css error mentioned above.

**Reply by AlessioGr** [reply]
This likely means that you are importing that custom component into your payload config somewhere, which is not allowed

**Reply by revnelson** [reply]
Then how can we render custom RowLabels? The documentation shows to use it this way: https://payloadcms.com/docs/fields/array#example-of-a-custom-rowlabel-component

**Reply by rilrom** [reply]
Are you importing the component into your config using this format?

```
admin: {
  components: {
    RowLabel: '/path/to/ArrayRowLabel#ArrayRowLabel',
  },
},
```

**Reply by revnelson** [reply]
Yes. The import itself works. Adding the `useRowLabel` hook auto-generates the correct importMap update and works fine. It's only when I run `payload generate:importMap` with `useRowLabel in the component that I get the error.

**Reply by rilrom** [reply]
If you can create a minimal reproduction repo we can help you resolve it.

**Reply by revnelson** [reply]
So it is allowed in the config? Is there a server-component way to get the data from props?

I don't have time to set up a reproduction repo because my config is complex and my work would not be happy with me posting the entire source code.

I started a new project with `create-payload-app` and the website template. I can see in the footer array [custom component](https://github.com/payloadcms/payload/blob/main/templates/website/src/Footer/RowLabel.tsx) that `useRowLabel` is working fine.

I think the main issue here is that the error being logged is entirely unhelpful for finding what the problem is. Without any useful debugging information I am stuck because the custom component works absolutely fine until I add `useRowLabel` (and removing it fixes the problem). This points to an issue with `useRowLabel`, but as I can see `useRowLabel` working correctly in the website template, I have nothing else to inspect.

Posting reproduction repos is a very high barrier, and an improvement in logging should be considered top priority with this major shift in importMap 'magic'.

**Reply by revnelson** [reply]
I've been playing with this more and the oddities continue. I wrote a util function to reduce boilerplate for custom component paths, as well as provide type safety for `clientProps`. Here is that util:

```ts
import type { RowLabelComponent } from "payload"
import type { ArrayRowLabelProps } from "./ArrayRowLabel"

export const AdminRowLabelConfig = (
	clientProps?: ArrayRowLabelProps
): RowLabelComponent => ({
	clientProps,
	path: "/components/ArrayRowLabel/ArrayRowLabel",
	exportName: "ArrayRowLabel"
})
```

And here is the custom component:
```ts
"use client"

import { useEffect, useState } from "react"
import { useRowLabel } from "@payloadcms/ui"

import { getRichText } from "~/utils/getRichText"

export type ArrayRowLabelProps = {
	labelKey?: string
	labelParent?: string
	fallback?: string
	employee?: boolean
	selectField?: string
}

export const ArrayRowLabel = ({
	labelKey = "label",
	labelParent,
	fallback = "Row",
	employee,
	selectField
}: ArrayRowLabelProps) => {
	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	const { data, rowNumber } = useRowLabel<{ [key: string]: any }>()

	const targetField =
		(labelParent && labelParent in data && data[labelParent]) || data

	const customLabelByKey =
		selectField && selectField in data
			? data[selectField]
					?.map((type: string) => `${type[0].toUpperCase()}${type.slice(1)}`)
					.join(" & ")
			: labelKey in targetField
				? targetField[labelKey]
				: null

	const customLabel =
		typeof customLabelByKey === "string"
			? customLabelByKey
			: getRichText(customLabelByKey)

	const [label, setLabel] = useState<string | null>(customLabel || null)

	const fetchEmployee = async (id: string) => {
		const response = await fetch(`/api/employees/${id}`)

		if (response.ok) {
			const employeeData = await response.json()

			if (employeeData.fullName) {
				setLabel(employeeData.fullName)
			}
		}
	}

	useEffect(() => {
		if (employee && targetField.employee) {
			fetchEmployee(targetField.employee)
		}
	}, [employee, targetField.employee])

	return (
		<div>{label || `${fallback} ${String(rowNumber).padStart(2, "0")} `}</div>
	)
}
```
----
Without changing any of the above, calling that utility in my config like `RowLabel: AdminRowLabelConfig({ fallback: "License" })` will cause the dreaded css issue when generating the importMap. However, manually entering the result of that util call allows the importMap to be generated without issue. Here's a global where it's used:

(NOT WORKING)
```ts
import { richTextField } from "~cms/fields/richText"
import { AdminRowLabelConfig } from "~cms/components/ArrayRowLabel"

import type { GlobalConfig } from "payload"

export const FooterSettings: GlobalConfig = {
	slug: "FooterSettings",
	label: "Footer",
	fields: [
		{
			name: "licenses",
			type: "array",
			fields: [
				richTextField({ name: "label" }),
				richTextField({ name: "license" })
			],
			admin: {
				components: {
					RowLabel: AdminRowLabelConfig({ fallback: "License" })
				}
			}
		}
	]
}
```

(WORKING)
```ts
import { richTextField } from "~cms/fields/richText"
// import { AdminRowLabelConfig } from "~cms/components/ArrayRowLabel"

import type { GlobalConfig } from "payload"

export const FooterSettings: GlobalConfig = {
	slug: "FooterSettings",
	label: "Footer",
	fields: [
		{
			name: "licenses",
			type: "array",
			fields: [
				richTextField({ name: "label" }),
				richTextField({ name: "license" })
			],
			admin: {
				components: {
					RowLabel: {
						clientProps: {
							fallback: "License"
						},
						path: "/components/ArrayRowLabel",
						exportName: "ArrayRowLabel"
					}
					// AdminRowLabelConfig({ fallback: "License" })
				}
			}
		}
	]
}
```

If I call that utility without putting it in my config, I also get the error:

```ts
import { richTextField } from "~cms/fields/richText"
import { AdminRowLabelConfig } from "~cms/components/ArrayRowLabel"

import type { GlobalConfig } from "payload"

const row = AdminRowLabelConfig({ fallback: "License" })

export const FooterSettings: GlobalConfig = {
	slug: "FooterSettings",
	label: "Footer",
	fields: [
		{
			name: "licenses",
			type: "array",
			fields: [
				richTextField({ name: "label" }),
				richTextField({ name: "license" })
			],
			admin: {
				components: {
					RowLabel: {
						clientProps: {
							fallback: "License"
						},
						path: "/components/ArrayRowLabel",
						exportName: "ArrayRowLabel"
					}
					// AdminRowLabelConfig({ fallback: "License" })
				}
			}
		}
	]
}
```

So it seems that something about that `AdminRowLabelConfig` util function is causing the issue. To troubleshoot a potential issue from the type being exported from the custom component and imported in the utility function, I copy/pasted it and removed the import/export but that didn't help anything.

**Reply by revnelson** [reply]
If I copy/paste the util into the global config file (rather than importing it), importMap generation works:

```ts
import { richTextField } from "~cms/fields/richText"

import type { GlobalConfig } from "payload"

import type { RowLabelComponent } from "payload"

type ArrayRowLabelProps = {
	labelKey?: string
	labelParent?: string
	fallback?: string
	employee?: boolean
	selectField?: string
}

const AdminRowLabelConfig = (
	clientProps?: ArrayRowLabelProps
): RowLabelComponent => ({
	clientProps,
	path: "/components/ArrayRowLabel",
	exportName: "ArrayRowLabel"
})

export const FooterSettings: GlobalConfig = {
	slug: "FooterSettings",
	label: "Footer",
	fields: [
		{
			name: "licenses",
			type: "array",
			fields: [
				richTextField({ name: "label" }),
				richTextField({ name: "license" })
			],
			admin: {
				components: {
					RowLabel: AdminRowLabelConfig({ fallback: "License" })
				}
			}
		}
	]
}
```

Do I need to add the path to that utility function somewhere to include/exclude from bundling somewhere?

**Comment by revnelson** [comment]
Enabling `verbatimModuleSyntax` in tsconfig also causes importMap generation to throw this error.
